File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:   
    def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:
        def same(first, second):
            if not first and not second:
                return True
            if not first or not second:
                return False
            if first.val != second.val:
                return False
            left = same(first.left, second.left)
            if not left:
                return False
            right = same(first.right, second.right)
            if not right:
                return False
            return left and right
        
        q = deque([root])
        while q:
            node = q.popleft()
            if same(node, subRoot):
                return True
            if node.left:
                q.append(node.left)
            if node.right:
                q.append(node.right)
        return False

      

File: Data Structures & Algorithms/binary-tree-diameter/submission-2.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {

private:
    int maxdiameter =0;

    int diameter (TreeNode* node){
            if (node == nullptr){
                return 0;
            }
            int left = diameter(node->left);
            int right = diameter(node->right);
            maxdiameter = max(maxdiameter, left+right);
            return 1+ max(left,right);

        }

public:

    int diameterOfBinaryTree(TreeNode* root) {

        diameter(root);
        return maxdiameter; 
    }

    

};
